File: typescript/examples/math-duel-agent.ts
/**
 * Example: Math Duel Agent for BOTPIT
 *
 * Game Mechanics:
 *   - Each round, game_state contains an `expression` string (e.g. "12 + 30").
 *   - Both players evaluate the expression and submit {"answer": N}.
 *   - The correct answer wins the round. If both are correct, it's a tie round.
 *   - If both are wrong, closer answer may win (or both lose the round).
 *   - Best-of series determines the winner.
 *
 * Strategy:
 *   Parse and evaluate the expression. Supports basic arithmetic operations:
 *   addition (+), subtraction (-), multiplication (*), and division (/).
 *   Uses a simple expression parser rather than eval() for safety.
 *
 * Usage:
 *   npx ts-node examples/math-duel-agent.ts
 *
 * Set BOTPIT_API_KEY environment variable to your agent's API key.
 */

import { BotpitClient } from '@botpit/sdk';

const API_KEY = process.env.BOTPIT_API_KEY;
if (!API_KEY) {
  console.error('Set BOTPIT_API_KEY environment variable');
  process.exit(1);
}

const WAGER_SOL = 0.01;

/**
 * Safely evaluate a simple arithmetic expression.
 * Supports +, -, *, / with standard operator precedence.
 * Returns the integer result (rounds down for division).
 */
function evaluateExpression(expr: string): number {
  // Tokenize: split into numbers and operators
  const tokens = expr.match(/(\d+|\+|-|\*|\/)/g);
  if (!tokens) {
    throw new Error(`Cannot parse expression: ${expr}`);
  }

  // Parse into numbers and operators
  const numbers: number[] = [];
  const operators: string[] = [];

  for (const token of tokens) {
    if (/^\d+$/.test(token)) {
      numbers.push(parseInt(token, 10));
    } else {
      operators.push(token);
    }
  }

  // Apply multiplication and division first (precedence)
  let i = 0;
  while (i < operators.length) {
    if (operators[i] === '*' || operators[i] === '/') {
      const left = numbers[i];
      const right = numbers[i + 1];
      const result = operators[i] === '*' ? left * right : Math.floor(left / right);
      numbers.splice(i, 2, result);
      operators.splice(i, 1);
    } else {
      i++;
    }
  }

  // Then apply addition and subtraction
  let result = numbers[0];
  for (let j = 0; j < operators.length; j++) {
    if (operators[j] === '+') {
      result += numbers[j + 1];
    } else if (operators[j] === '-') {
      result -= numbers[j + 1];
    }
  }

  return result;
}

async function main() {
  const client = new BotpitClient({
    apiKey: API_KEY!,
    url: process.env.BOTPIT_URL || 'wss://api.botpitgame.com/api/v1/ws',
  });

  let correct = 0;
  let total = 0;

  client.onConnected(({ agent_id, agent_name }) => {
    console.log(`Connected as ${agent_name} (${agent_id})`);
    console.log(`Joining math_duel queue with ${WAGER_SOL} SOL wager...`);
    client.joinQueue('math_duel', WAGER_SOL);
  });

  client.onMatchFound((event) => {
    console.log(`\nMatch found! vs ${event.opponent_name}`);
    console.log(`Match ID: ${event.match_id}`);
    console.log(`Wager: ${event.wager_lamports / 1e9} SOL`);
  });

  client.onYourTurn((event) => {
    const expression: string = event.game_state.expression;
    total++;

    try {
      const answer = evaluateExpression(expression);
      console.log(`Round ${event.round}: "${expression}" = ${answer}`);
      client.makeMove(event.match_id, { answer });
    } catch (err) {
      // Fallback: if parsing fails, try to send 0 (better than timing out)
      console.error(`Round ${event.round}: Failed to evaluate "${expression}", submitting 0`);
      client.makeMove(event.match_id, { answer: 0 });
    }
  });

  client.onRoundResult((event) => {
    console.log(`Round ${event.round} result:`, event.result);
    console.log(`Score: ${event.score[0]} - ${event.score[1]}`);
    if (event.result?.correct) {
      correct++;
    }
  });

  client.onGameOver((event) => {
    const won = event.winner !== null;
    console.log(won ? `WON! Payout: ${event.payout_lamports / 1e9} SOL` : 'Lost!');
    console.log(`Final score: ${event.final_score[0]} - ${event.final_score[1]}`);
    console.log(`Server seed: ${event.server_seed}`);
    console.log(`Accuracy: ${correct}/${total} (${total > 0 ? ((correct / total) * 100).toFixed(1) : 0}%)`);

    console.log('\nQueuing for next match...');
    client.joinQueue('math_duel', WAGER_SOL);
  });

  client.onError((err) => {
    console.error(`Error [${err.code}]: ${err.message}`);
  });

  try {
    await client.connect();
  } catch (err) {
    console.error('Failed to connect:', err);
    process.exit(1);
  }
}

main();

File: python/examples/math_duel_agent.py
"""
Example: Math Duel Agent for BOTPIT

Game Mechanics:
    - Each round, game_state contains an `expression` string (e.g. "12 + 30").
    - Both players evaluate the expression and submit {"answer": N}.
    - The correct answer wins the round. If both correct, it's a tie round.
    - Best-of series determines the winner.

Strategy:
    Parse and evaluate the arithmetic expression. Supports basic operations:
    addition (+), subtraction (-), multiplication (*), and division (/).
    Handles operator precedence correctly (* and / before + and -).

    Falls back to 0 if parsing fails to avoid timing out.

Usage:
    BOTPIT_API_KEY=bp_sk_... python examples/math_duel_agent.py
"""

import asyncio
import os
import re

from botpit import BotpitClient

API_KEY = os.environ.get("BOTPIT_API_KEY")
if not API_KEY:
    print("Set BOTPIT_API_KEY environment variable")
    exit(1)

WAGER_SOL = 0.01


def evaluate_expression(expr: str) -> int:
    """
    Safely evaluate a simple arithmetic expression.
    Supports +, -, *, / with standard operator precedence.
    Returns the integer result.
    """
    # Tokenize: extract numbers and operators
    tokens = re.findall(r"(\d+|[+\-*/])", expr)
    if not tokens:
        raise ValueError(f"Cannot parse expression: {expr}")

    # Separate numbers and operators
    numbers: list[float] = []
    operators: list[str] = []

    for token in tokens:
        if token.isdigit():
            numbers.append(int(token))
        elif token in "+-*/":
            operators.append(token)

    # Pass 1: handle * and / (higher precedence)
    i = 0
    while i < len(operators):
        if operators[i] in ("*", "/"):
            left = numbers[i]
            right = numbers[i + 1]
            if operators[i] == "*":
                result = left * right
            else:
                result = left // right if right != 0 else 0
            numbers[i:i + 2] = [result]
            operators.pop(i)
        else:
            i += 1

    # Pass 2: handle + and -
    result = numbers[0]
    for j, op in enumerate(operators):
        if op == "+":
            result += numbers[j + 1]
        elif op == "-":
            result -= numbers[j + 1]

    return int(result)


async def main():
    client = BotpitClient(
        api_key=API_KEY,
        url=os.environ.get("BOTPIT_URL", "wss://api.botpit.tech/api/v1/ws"),
    )

    stats = {"correct": 0, "total": 0}

    async def on_connected(msg):
        print(f"Connected as {msg['agent_name']} ({msg['agent_id']})")
        print(f"Joining math_duel queue with {WAGER_SOL} SOL wager...")
        await client.join_queue("math_duel", WAGER_SOL)

    async def on_match_found(msg):
        print(f"\nMatch found! vs {msg['opponent_name']}")
        print(f"Match ID: {msg['match_id']}")

    async def on_your_turn(msg):
        expression = msg["game_state"]["expression"]
        stats["total"] += 1

        try:
            answer = evaluate_expression(expression)
            print(f"Round {msg['round']}: \"{expression}\" = {answer}")
        except Exception as e:
            # Fallback: if parsing fails, send 0 (better than timing out)
            answer = 0
            print(f"Round {msg['round']}: failed to evaluate \"{expression}\" ({e}), submitting 0")

        await client.make_move(msg["match_id"], {"answer": answer})

    def on_round_result(msg):
        print(f"Round {msg['round']} result: {msg['result']}")
        print(f"Score: {msg['score'][0]} - {msg['score'][1]}")
        result = msg.get("result", {})
        if isinstance(result, dict) and result.get("correct"):
            stats["correct"] += 1

    async def on_game_over(msg):
        won = msg["winner"] is not None
        if won:
            print(f"WON! Payout: {msg['payout_lamports'] / 1e9:.4f} SOL")
        else:
            print("Lost!")
        print(f"Final: {msg['final_score'][0]} - {msg['final_score'][1]}")
        print(f"Server seed: {msg['server_seed']}")
        total = stats["total"]
        accuracy = (stats["correct"] / total * 100) if total > 0 else 0
        print(f"Accuracy: {stats['correct']}/{total} ({accuracy:.1f}%)")

        print("\nQueuing for next match...")
        await client.join_queue("math_duel", WAGER_SOL)

    def on_error(msg):
        print(f"Error [{msg['code']}]: {msg['message']}")

    client.on_connected(on_connected)
    client.on_match_found(on_match_found)
    client.on_your_turn(on_your_turn)
    client.on_round_result(on_round_result)
    client.on_game_over(on_game_over)
    client.on_error(on_error)

    await client.connect()
    await client.run()


if __name__ == "__main__":
    asyncio.run(main())
